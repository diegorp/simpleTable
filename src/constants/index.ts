export const PAGE_SIZES = [5, 10, 25, 50, 100];

export const ENTER_KEY = 'Enter';

export const ARIA_SORT_VALUES = {
  asc: 'ascending',
  desc: 'descending',
  none: 'other',
} as const;
