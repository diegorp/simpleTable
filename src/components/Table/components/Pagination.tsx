import { PAGE_SIZES } from '@/constants';

interface Props {
  currentPage: number;
  currentPageSize: number;
  totalItems: number;
  isLastPage: boolean;
  onPageChange: (page: number) => void;
  onPageSizeChange: (pageSize: number) => void;
}

export default function Pagination({
  currentPage,
  currentPageSize,
  totalItems,
  isLastPage,
  onPageChange,
  onPageSizeChange,
}: Props) {
  return (
    <nav
      className='flex shrink-0 gap-12 w-full text-sm text-gray-500 items-center justify-start'
      aria-label='Table pagination'
    >
      <div className='flex items-center gap-2'>
        <span>Rows per page: </span>
        <select
          onChange={({ currentTarget }) =>
            onPageSizeChange(Number(currentTarget?.value ?? currentPageSize))
          }
          aria-label='Rows per page'
          aria-valuenow={currentPageSize}
        >
          {PAGE_SIZES.map(pageSizeOption => (
            <option key={pageSizeOption} value={pageSizeOption}>
              {pageSizeOption}
            </option>
          ))}
        </select>
      </div>
      <div className='flex items-center gap-2'>
        <span role='status'>
          {currentPage * currentPageSize + 1} -{' '}
          {Math.min((currentPage + 1) * currentPageSize, totalItems)} of{' '}
          {totalItems}
        </span>
        <button
          className='text-xl cursor-pointer mr-2 disabled:opacity-50 disabled:cursor-not-allowed'
          onClick={() => onPageChange(currentPage - 1)}
          disabled={currentPage === 0}
          aria-label='Previous page'
        >
          ‹
        </button>
        <button
          className='text-xl cursor-pointer disabled:opacity-50 disabled:cursor-not-allowed'
          onClick={() => onPageChange(currentPage + 1)}
          disabled={isLastPage}
          aria-label='Next page'
        >
          ›
        </button>
      </div>
    </nav>
  );
}
