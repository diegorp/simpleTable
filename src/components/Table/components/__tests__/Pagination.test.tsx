import { render, fireEvent } from '@testing-library/react';
import Pagination from '@/components/Table/components/Pagination';
import { PAGE_SIZES } from '@/constants';

describe('<Pagination />', () => {
  const defaultProps = {
    currentPage: 0,
    currentPageSize: 10,
    totalItems: 50,
    isLastPage: false,
    onPageChange: vi.fn(),
    onPageSizeChange: vi.fn(),
  };

  it('should match snapshot', () => {
    const { container } = render(<Pagination {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it('should display current page range', () => {
    const { getByText } = render(<Pagination {...defaultProps} />);

    expect(getByText(/1 - 10 of 50/)).toBeInTheDocument();
  });

  it('should display correct range for second page', () => {
    const { getByText } = render(
      <Pagination {...defaultProps} currentPage={1} />
    );

    expect(getByText(/11 - 20 of 50/)).toBeInTheDocument();
  });

  it('should display correct range for last page', () => {
    const { getByText } = render(
      <Pagination
        {...defaultProps}
        currentPage={4}
        totalItems={45}
        isLastPage={true}
      />
    );

    expect(getByText(/41 - 45 of 45/)).toBeInTheDocument();
  });

  it('should calculate correct range with different page size', () => {
    const { getByText } = render(
      <Pagination {...defaultProps} currentPage={1} currentPageSize={25} />
    );

    expect(getByText('26 - 50 of 50')).toBeInTheDocument();
  });

  it('should render page size selector', () => {
    const { getByText, getByRole } = render(<Pagination {...defaultProps} />);

    expect(getByText('Rows per page:')).toBeInTheDocument();
    expect(
      getByRole('combobox', { name: 'Rows per page' })
    ).toBeInTheDocument();
  });

  it('should render previous and next buttons', () => {
    const { getAllByRole } = render(<Pagination {...defaultProps} />);

    const buttons = getAllByRole('button');
    expect(buttons).toHaveLength(2);
    expect(buttons[0]).toHaveTextContent('‹');
    expect(buttons[1]).toHaveTextContent('›');
  });

  it('should disable previous button on first page', () => {
    const { getByRole } = render(
      <Pagination {...defaultProps} currentPage={0} />
    );

    const prevButton = getByRole('button', { name: 'Previous page' });
    expect(prevButton).toBeDisabled();
  });

  it('should enable previous button when not on first page', () => {
    const { getByRole } = render(
      <Pagination {...defaultProps} currentPage={1} />
    );

    const prevButton = getByRole('button', { name: 'Previous page' });
    expect(prevButton).not.toBeDisabled();
  });

  it('should disable next button on last page', () => {
    const { getByRole } = render(
      <Pagination {...defaultProps} isLastPage={true} />
    );

    const nextButton = getByRole('button', { name: 'Next page' });
    expect(nextButton).toBeDisabled();
  });

  it('should enable next button when not on last page', () => {
    const { getByRole } = render(
      <Pagination {...defaultProps} isLastPage={false} />
    );

    const nextButton = getByRole('button', { name: 'Next page' });
    expect(nextButton).not.toBeDisabled();
  });

  it('should call onPageChange when previous button is clicked', () => {
    const onPageChange = vi.fn();
    const { getByRole } = render(
      <Pagination
        {...defaultProps}
        currentPage={1}
        onPageChange={onPageChange}
      />
    );

    const prevButton = getByRole('button', { name: 'Previous page' });
    fireEvent.click(prevButton);

    expect(onPageChange).toHaveBeenCalledWith(0);
  });

  it('should call onPageChange when next button is clicked', () => {
    const onPageChange = vi.fn();
    const { getByRole } = render(
      <Pagination {...defaultProps} onPageChange={onPageChange} />
    );

    const nextButton = getByRole('button', { name: 'Next page' });
    fireEvent.click(nextButton);

    expect(onPageChange).toHaveBeenCalledWith(1);
  });

  it('should call onPageSizeChange when page size is changed', () => {
    const onPageSizeChange = vi.fn();
    const { getByRole } = render(
      <Pagination {...defaultProps} onPageSizeChange={onPageSizeChange} />
    );

    const select = getByRole('combobox', { name: 'Rows per page' });
    fireEvent.change(select, { target: { value: '25' } });

    expect(onPageSizeChange).toHaveBeenCalledWith(25);
  });

  it('should display available page size options', () => {
    const { getByRole } = render(<Pagination {...defaultProps} />);

    const select = getByRole('combobox', { name: 'Rows per page' });
    const options = select.querySelectorAll('option');

    expect(options.length).toEqual(PAGE_SIZES.length);
  });

  it('should handle single page scenario', () => {
    const { getByText, getByRole } = render(
      <Pagination
        currentPage={0}
        currentPageSize={50}
        totalItems={10}
        isLastPage={true}
        onPageChange={vi.fn()}
        onPageSizeChange={vi.fn()}
      />
    );

    expect(getByText(/1 - 10 of 10/)).toBeInTheDocument();
    expect(getByRole('button', { name: 'Previous page' })).toBeDisabled();
    expect(getByRole('button', { name: 'Next page' })).toBeDisabled();
  });

  it('should handle empty data', () => {
    const { getByText } = render(
      <Pagination
        currentPage={0}
        currentPageSize={10}
        totalItems={0}
        isLastPage={true}
        onPageChange={vi.fn()}
        onPageSizeChange={vi.fn()}
      />
    );

    expect(getByText(/1 - 0 of 0/)).toBeInTheDocument();
  });
});
