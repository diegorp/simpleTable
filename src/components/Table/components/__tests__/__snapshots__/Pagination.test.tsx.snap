// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Pagination /> > should match snapshot 1`] = `
<div>
  <nav
    aria-label="Table pagination"
    class="flex shrink-0 gap-12 w-full text-sm text-gray-500 items-center justify-start"
  >
    <div
      class="flex items-center gap-2"
    >
      <span>
        Rows per page: 
      </span>
      <select
        aria-label="Rows per page"
        aria-valuenow="10"
      >
        <option
          value="5"
        >
          5
        </option>
        <option
          value="10"
        >
          10
        </option>
        <option
          value="25"
        >
          25
        </option>
        <option
          value="50"
        >
          50
        </option>
        <option
          value="100"
        >
          100
        </option>
      </select>
    </div>
    <div
      class="flex items-center gap-2"
    >
      <span
        role="status"
      >
        1
         -
         
        10
         of
         
        50
      </span>
      <button
        aria-label="Previous page"
        class="text-xl cursor-pointer mr-2 disabled:opacity-50 disabled:cursor-not-allowed"
        disabled=""
      >
        ‹
      </button>
      <button
        aria-label="Next page"
        class="text-xl cursor-pointer disabled:opacity-50 disabled:cursor-not-allowed"
      >
        ›
      </button>
    </div>
  </nav>
</div>
`;
